add administrative capabilities and best practices of system level Groovy (#8038)
jenkins-infra/jenkins.io: content/doc/book/managing/script-console.adoc

interface for Jenkins developers but has since grown into an interface used by
Jenkins Admins to configure Jenkins and debug Jenkins runtime issues.

Additional security considerations for administrative scripts:

> * Scripts can bypass all security controls and access any data

Reviewer: ```suggestion
* Scripts can bypass all security controls and access any data.
```
Adding full stop at the end of the sentence.

> * Credentials accessed via scripts are decrypted

Reviewer: ```suggestion
* Credentials accessed via scripts are decrypted.
```
Same as above, adding full stop.

> * Changes made via scripts may not be audited in the normal logs

Reviewer: ```suggestion
* Changes made via scripts may not be audited in the normal logs.
```
Adding full stop.

> * Scripts can modify Jenkins at the lowest levels, potentially making it unstable

Reviewer: ```suggestion
* Scripts can modify Jenkins at the lowest levels, potentially making it unstable.
```
Adding full stop

println '==='
----

== Administrative capabilities

The Script Console provides powerful administrative functions for managing Jenkins.

Below are common administrative tasks with examples:

=== User management
[source,groovy]
----
Jenkins.instance.securityRealm.allUsers.each { user ->
    println user.id + ": " + user.fullName
}

import hudson.model.User
import jenkins.model.Jenkins

User user = User.get('new-user', true)
user.fullName = "New User"
user.save()
----

=== System configuration
[source,groovy]
----
Jenkins.instance.systemMessage = "New system message"
Jenkins.instance.save()

println Jenkins.VERSION
----

=== Plugin management
[source,groovy]
----
Jenkins.instance.pluginManager.plugins.each {
    println "${it.shortName}: ${it.version}"
}
Jenkins.instance.pluginManager.getPlugin('git').disable()
----

=== Node/Agent management
[source,groovy]
----
Jenkins.instance.nodes.each { node ->
    println "${node.name}: ${node.numExecutors} executors"
}
Jenkins.instance.getNode('agent-name').computer.doDoDelete()
----

== Best practices for system groovy

When using the Script Console for administrative tasks:

> 1. *Always test scripts in a non-production environment first*
> 2. *Backup your Jenkins configuration* before running scripts
> 3. *Use the Jenkins API* rather than direct filesystem manipulation when possible
> 4. *Limit script scope* - only perform necessary operations
> 5. *Add comments* to scripts for future reference
> 6. *Consider using plugins* like Job DSL or Configuration as Code for repeatable configurations
> 7. *Review scripts carefully* - they execute with full system privileges
> 8. *Monitor script execution* - long-running scripts may impact performance

Reviewer: ```suggestion
* Always test scripts in a non-production environment first.
* link:/doc/book/system-administration/backing-up/[Backup your Jenkins configuration] before running scripts.
* Use the link:/doc/developer/security/misc/[Jenkins API] rather than direct filesystem manipulation when possible.
* Limit script scope so that only necessary operations are performed.
* Add comments to scripts for future reference.
* Consider using plugins like plugin:job-dsl[Job DSL] or plugin:configuration-as-code[Configuration as Code] for repeatable configurations.
* Review scripts carefully to ensure they execute with full system privileges.
* Monitor script execution as long-running scripts may impact performance.
```
Suggestion to replace the numbered list with bullets as these are not steps/should not be ranked. I've also turned them all into full sentences and included the links for corresponding plugins & existing documentation.
